File: code_solutions.js

// JavaScript Solutions

function reverseWords(sentence) {
  return sentence.trim().split(/\s+/).reverse().join(" ");
}

function rotateArray(arr, k) {
  k = k % arr.length;
  return arr.slice(-k).concat(arr.slice(0, -k));
}

function isValidParentheses(s) {
  const stack = [];
  const map = {')': '(', ']': '[', '}': '{'};
  for (let char of s) {
    if (['(', '[', '{'].includes(char)) stack.push(char);
    else if (stack.pop() !== map[char]) return false;
  }
  return stack.length === 0;
}

function longestCommonPrefix(strs) {
  if (!strs.length) return "";
  let prefix = strs[0];
  for (let i = 1; i < strs.length; i++) {
    while (strs[i].indexOf(prefix) !== 0) {
      prefix = prefix.slice(0, -1);
      if (!prefix) return "";
    }
  }
  return prefix;
}

function fizzBuzz(n) {
  for (let i = 1; i <= n; i++) {
    if (i % 15 === 0) console.log("FizzBuzz");
    else if (i % 3 === 0) console.log("Fizz");
    else if (i % 5 === 0) console.log("Buzz");
    else console.log(i);
  }
}

function hasUniqueChars(str) {
  const set = new Set();
  for (let char of str) {
    if (set.has(char)) return false;
    set.add(char);
  }
  return true;
}

function twoSum(nums, target) {
  const map = new Map();
  for (let i = 0; i < nums.length; i++) {
    let complement = target - nums[i];
    if (map.has(complement)) return [map.get(complement), i];
    map.set(nums[i], i);
  }
  return [];
}

// Import readline for user input in Node.js
const readline = require('readline');

// Create readline interface
const rl = readline.createInterface({
  input: process.stdin,
  output: process.stdout
});

// Function to ask questions and return promises
function askQuestion(question) {
  return new Promise((resolve) => {
    rl.question(question, (answer) => {
      resolve(answer);
    });
  });
}

async function main() {
  console.log("Welcome to JavaScript Code Solutions!");
  
  while (true) {
    console.log("\n=== JavaScript Code Solutions Menu ===");
    console.log("1. Reverse Words in a Sentence");
    console.log("2. Rotate Array");
    console.log("3. Validate Parentheses");
    console.log("4. Find Longest Common Prefix");
    console.log("5. FizzBuzz Game");
    console.log("6. Check if String has Unique Characters");
    console.log("7. Two Sum Problem");
    console.log("0. Exit");
    
    try {
      const choice = await askQuestion("Enter your choice (0-7): ");
      
      switch (parseInt(choice)) {
        case 1:
          const sentence = await askQuestion("Enter a sentence to reverse words: ");
          const reversed = reverseWords(sentence);
          console.log(`Original: ${sentence}`);
          console.log(`Reversed: ${reversed}`);
          break;
          
        case 2:
          const arrInput = await askQuestion("Enter array elements separated by commas: ");
          const arr = arrInput.split(',').map(num => parseInt(num.trim()));
          const k = parseInt(await askQuestion("Enter rotation count: "));
          const rotated = rotateArray(arr, k);
          console.log(`Original array: [${arr.join(', ')}]`);
          console.log(`Rotated by ${k}: [${rotated.join(', ')}]`);
          break;
          
        case 3:
          const parentheses = await askQuestion("Enter string with parentheses to validate: ");
          const isValid = isValidParentheses(parentheses);
          console.log(`'${parentheses}' has ${isValid ? 'valid' : 'invalid'} parentheses`);
          break;
          
        case 4:
          const strsInput = await askQuestion("Enter strings separated by commas: ");
          const strs = strsInput.split(',').map(str => str.trim());
          const prefix = longestCommonPrefix(strs);
          console.log(`Strings: [${strs.map(s => `"${s}"`).join(', ')}]`);
          console.log(`Longest common prefix: "${prefix}"`);
          break;
          
        case 5:
          const n = parseInt(await askQuestion("Enter number for FizzBuzz game: "));
          console.log("FizzBuzz output:");
          fizzBuzz(n);
          break;
          
        case 6:
          const uniqueStr = await askQuestion("Enter string to check for unique characters: ");
          const hasUnique = hasUniqueChars(uniqueStr);
          console.log(`'${uniqueStr}' ${hasUnique ? 'has all unique characters' : 'has duplicate characters'}`);
          break;
          
        case 7:
          const numsInput = await askQuestion("Enter array elements separated by commas: ");
          const nums = numsInput.split(',').map(num => parseInt(num.trim()));
          const target = parseInt(await askQuestion("Enter target sum: "));
          const result = twoSum(nums, target);
          if (result.length === 2) {
            console.log(`Indices: [${result[0]}, ${result[1]}]`);
            console.log(`Values: [${nums[result[0]]}, ${nums[result[1]]}]`);
          } else {
            console.log("No two numbers sum to target");
          }
          break;
          
        case 0:
          console.log("Goodbye! Thanks for using JavaScript Code Solutions!");
          rl.close();
          return;
          
        default:
          console.log("Invalid choice! Please enter a number between 0-7.");
          break;
      }
      
      await askQuestion("\nPress Enter to continue...");
      
    } catch (error) {
      console.log("An error occurred:", error.message);
    }
  }
}

// Run the main function if this file is executed directly
if (require.main === module) {
  main().catch(console.error);
}

File: CodeSolutions.java

// Java Solutions
import java.util.*;

public class CodeSolutions {

    public static String reverseWords(String sentence) {
        String[] words = sentence.trim().split("\\s+");
        Collections.reverse(Arrays.asList(words));
        return String.join(" ", words);
    }

    public static boolean isAnagram(String s1, String s2) {
        char[] a = s1.replaceAll("\\s", "").toLowerCase().toCharArray();
        char[] b = s2.replaceAll("\\s", "").toLowerCase().toCharArray();
        Arrays.sort(a);
        Arrays.sort(b);
        return Arrays.equals(a, b);
    }

    public static int findMissing(int[] nums) {
        int n = nums.length + 1;
        int total = n * (n + 1) / 2;
        for (int num : nums) total -= num;
        return total;
    }

    public static boolean isValidParentheses(String s) {
        Stack<Character> stack = new Stack<>();
        for (char c : s.toCharArray()) {
            if (c == '(' || c == '[' || c == '{') stack.push(c);
            else if (c == ')' && !stack.isEmpty() && stack.peek() == '(') stack.pop();
            else if (c == ']' && !stack.isEmpty() && stack.peek() == '[') stack.pop();
            else if (c == '}' && !stack.isEmpty() && stack.peek() == '{') stack.pop();
            else return false;
        }
        return stack.isEmpty();
    }

    public static void fizzBuzz(int n) {
        for (int i = 1; i <= n; i++) {
            if (i % 15 == 0) System.out.println("FizzBuzz");
            else if (i % 3 == 0) System.out.println("Fizz");
            else if (i % 5 == 0) System.out.println("Buzz");
            else System.out.println(i);
        }
    }

    public static boolean hasUniqueChars(String s) {
        Set<Character> set = new HashSet<>();
        for (char c : s.toCharArray()) {
            if (set.contains(c)) return false;
            set.add(c);
        }
        return true;
    }

    public static int[] twoSum(int[] nums, int target) {
        Map<Integer, Integer> map = new HashMap<>();
        for (int i = 0; i < nums.length; i++) {
            int complement = target - nums[i];
            if (map.containsKey(complement)) return new int[]{map.get(complement), i};
            map.put(nums[i], i);
        }
        return new int[]{};
    }

    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        
        while (true) {
            System.out.println("\n=== Java Code Solutions Menu ===");
            System.out.println("1. Reverse Words in a Sentence");
            System.out.println("2. Check if Two Strings are Anagrams");
            System.out.println("3. Find Missing Number in Array");
            System.out.println("4. Validate Parentheses");
            System.out.println("5. FizzBuzz Game");
            System.out.println("6. Check if String has Unique Characters");
            System.out.println("7. Two Sum Problem");
            System.out.println("0. Exit");
            System.out.print("Enter your choice (0-7): ");
            
            int choice = scanner.nextInt();
            scanner.nextLine(); // consume newline
            
            switch (choice) {
                case 1:
                    System.out.print("Enter a sentence to reverse words: ");
                    String sentence = scanner.nextLine();
                    String reversed = reverseWords(sentence);
                    System.out.println("Original: " + sentence);
                    System.out.println("Reversed: " + reversed);
                    break;
                    
                case 2:
                    System.out.print("Enter first string: ");
                    String str1 = scanner.nextLine();
                    System.out.print("Enter second string: ");
                    String str2 = scanner.nextLine();
                    boolean isAnag = isAnagram(str1, str2);
                    System.out.println("'" + str1 + "' and '" + str2 + "' are " + 
                                     (isAnag ? "anagrams" : "not anagrams"));
                    break;
                    
                case 3:
                    System.out.print("Enter array size (n): ");
                    int n = scanner.nextInt();
                    int[] nums = new int[n-1];
                    System.out.println("Enter " + (n-1) + " numbers (1 to " + n + " with one missing): ");
                    for (int i = 0; i < nums.length; i++) {
                        nums[i] = scanner.nextInt();
                    }
                    int missing = findMissing(nums);
                    System.out.println("Missing number is: " + missing);
                    break;
                    
                case 4:
                    System.out.print("Enter string with parentheses to validate: ");
                    String parentheses = scanner.nextLine();
                    boolean isValid = isValidParentheses(parentheses);
                    System.out.println("'" + parentheses + "' has " + 
                                     (isValid ? "valid" : "invalid") + " parentheses");
                    break;
                    
                case 5:
                    System.out.print("Enter number for FizzBuzz game: ");
                    int fizzBuzzN = scanner.nextInt();
                    System.out.println("FizzBuzz output:");
                    fizzBuzz(fizzBuzzN);
                    break;
                    
                case 6:
                    System.out.print("Enter string to check for unique characters: ");
                    String uniqueStr = scanner.nextLine();
                    boolean hasUnique = hasUniqueChars(uniqueStr);
                    System.out.println("'" + uniqueStr + "' " + 
                                     (hasUnique ? "has all unique characters" : "has duplicate characters"));
                    break;
                    
                case 7:
                    System.out.print("Enter array size: ");
                    int arrSize = scanner.nextInt();
                    int[] twoSumArray = new int[arrSize];
                    System.out.println("Enter " + arrSize + " numbers: ");
                    for (int i = 0; i < arrSize; i++) {
                        twoSumArray[i] = scanner.nextInt();
                    }
                    System.out.print("Enter target sum: ");
                    int target = scanner.nextInt();
                    int[] result = twoSum(twoSumArray, target);
                    if (result.length == 2) {
                        System.out.println("Indices: [" + result[0] + ", " + result[1] + "]");
                        System.out.println("Values: [" + twoSumArray[result[0]] + ", " + twoSumArray[result[1]] + "]");
                    } else {
                        System.out.println("No two numbers sum to target");
                    }
                    break;
                    
                case 0:
                    System.out.println("Goodbye! Thanks for using Code Solutions!");
                    scanner.close();
                    return;
                    
                default:
                    System.out.println("Invalid choice! Please enter a number between 0-7.");
                    break;
            }
            
            System.out.println("\nPress Enter to continue...");
            scanner.nextLine();
        }
    }
}
